src/SFML/Graphics/SpriteBatch.cpp

```diff
@@ -0,0 +1,267 @@
+{
+    if (m_updateRequired)
+    {
+        updateBatch();
+        m_updateRequired = false;
+    }
+
+    RenderStates statesCopy(states);
+
+    std::size_t startTriangle = 0;
+
+    for (std::size_t i = 0; i < m_batches.size(); i++)
+    {
+        statesCopy.texture = m_batches[i].texture;
+
+        const auto& vertices = m_batchMode == BatchMode::Deferred ? m_unsortedVertices : m_vertices;
+
+        target.draw(&vertices[startTriangle], m_batches[i].vertexCount, PrimitiveType::Triangles, statesCopy);
+
+        startTriangle += m_batches[i].vertexCount;
+    }
+}
+
+
+////////////////////////////////////////////////////////////
+void SpriteBatch::pushTriangle(const Vertex&    a,
+                               const Vertex&    b,
+                               const Vertex&    c,
+                               const Transform& transform,
+                               const Texture*   texture,
+                               float            depth)
+{
+    m_triangles.emplace_back(texture, depth);
+
+    m_unsortedVertices.emplace_back(transform * a.position, a.color, a.texCoords);
+    m_unsortedVertices.emplace_back(transform * b.position, b.color, b.texCoords);
+    m_unsortedVertices.emplace_back(transform * c.position, c.color, c.texCoords);
+
+    m_updateRequired = true;
+}
+
+
+////////////////////////////////////////////////////////////
+void SpriteBatch::updateBatch() const
+{
+    // Sanity check
+    if (m_triangles.size() == 0)
+        return;
+
+    if (m_batchMode == BatchMode::Deferred)
+    {
+        // Batch based on texture change
+        std::size_t    startIndex  = 0;
+        std::size_t    nextIndex   = 0;
+        const Texture* lastTexture = nullptr;
+
+        while (nextIndex < m_triangles.size())
+        {
+            const Texture* nextTexture = m_triangles[nextIndex].texture;
+            if (nextTexture != lastTexture)
+            {
+                m_batches.emplace_back(lastTexture, (nextIndex - startIndex) * 3);
+                lastTexture = nextTexture;
+                startIndex  = nextIndex;
+            }
+
+            nextIndex++;
+        }
+
+        // Deal with leftovers
+        if (startIndex != m_triangles.size())
+        {
+            m_batches.emplace_back(lastTexture, (m_triangles.size() - startIndex) * 3);
+        }
+    }
+
+    else if (m_batchMode == BatchMode::TextureSort || m_batchMode == BatchMode::DepthSort)
+    {
+        std::vector<std::size_t> indices(m_triangles.size());
+
+        for (std::size_t i = 0; i < m_triangles.size(); i++)
+            indices[i] = i;
+
+        if (m_batchMode == BatchMode::TextureSort)
+        {
+            const auto comp = [&](const std::size_t a, const std::size_t b)
+            { return m_triangles[a].texture < m_triangles[b].texture; };
+
+            std::stable_sort(indices.begin(), indices.end(), comp);
+        }
+
+        else
+        {
+            const auto comp = [&](const std::size_t a, const std::size_t b)
```

Comment:
Instead of using this default capture mode (`[&]`) is it feasible to list the things you want to capture?

Reply:
Would capturing `this` work?

Reply:
No because that's the other default capture mode. Try to explicitly list the objects you're capturing. Sometimes that is not feasible but at least try to make it work as to minimize how much stuff gets put into the scope of the lambda.

Reply:
I think capturing `m_triangles` directly won't work, but I can do something like this:
```cpp
auto&      triangles  = this->m_triangles;
const auto comp      = [triangles](const std::size_t a, const std::size_t b)
{
    if (triangles[a].depth != triangles[b].depth)
        return triangles[a].depth > triangles[b].depth;
    else
        return triangles[a].texture < triangles[b].texture;
};
```

Reply:
Hmm, maybe you're right to use `[this]`. You can copy `m_triangles` into the lambda but that's wasteful. I don't hate the idea of making a local reference and using that instead. In either case, it's more explicit than using `[&]` which will also capture everything in the current scope.

Reply:
It seems like it's possible to capture `m_triangles` with an initializer such as `[&triangles = m_triangles]`. Maybe I'll use that instead.

Reply:
I think that's just making a non-const copy.